apiserver/facades/client/charms/services/repofactory.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package services
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+
+	"github.com/juju/juju/charmhub"
+	corecharm "github.com/juju/juju/core/charm"
+	charmrepo "github.com/juju/juju/core/charm/repository"
+)
+
+// CharmRepoFactoryConfig encapsulates the information required for creating a
+// new CharmRepoFactory instance.
+type CharmRepoFactoryConfig struct {
+	// The logger to use.
+	Logger loggo.Logger
+
+	// A transport that is injected when making charmhub API calls.
+	Transport charmhub.Transport
+
+	StateBackend StateBackend
+	ModelBackend ModelBackend
+}
+
+// CharmRepoFactory instantitates charm repositories. It memoizes results
+type CharmRepoFactory struct {
+	logger            loggo.Logger
+	charmhubTransport charmhub.Transport
+	stateBackend      StateBackend
+	modelBackend      ModelBackend
+
+	mu            sync.Mutex
+	memoizedRepos map[corecharm.Source]corecharm.Repository
+}
+
+// NewCharmRepoFactory returns a new factory instance with the provided configuration.
+func NewCharmRepoFactory(cfg CharmRepoFactoryConfig) *CharmRepoFactory {
+	return &CharmRepoFactory{
+		logger:            cfg.Logger,
+		charmhubTransport: cfg.Transport,
+		stateBackend:      cfg.StateBackend,
+		modelBackend:      cfg.ModelBackend,
+		memoizedRepos:     make(map[corecharm.Source]corecharm.Repository),
+	}
+}
+
+// GetCharmRepository returns a suitable corecharm.Repository instance for the
+// requested source. Lookups are memoized for future requests.
+func (f *CharmRepoFactory) GetCharmRepository(src corecharm.Source) (corecharm.Repository, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if repo, isCached := f.memoizedRepos[src]; isCached {
```

Comment:
This was a TODO in charms/client.go. This store memoizes repositories so we can reuse them in facade calls that deal with multiple entities.